microbial_thermodynamics/cell_growth_ODE.py

```diff
@@ -94,22 +94,171 @@ def dr(tr: float, a: float, R: float) -> float:
+    """
+    E = calculate_E(v, R)
+    theta = calculate_theta(reaction_energy, s, w)
+    u = (ka * E * s * (1 - theta)) / (ks + s * (1 + ra * theta))
+    return u
+
+
+def calculate_j(
+    R: float,
+    s: float,
+    w: float,
+    v: float,
+    reaction_energy: float,
+    Gatp: float = 75000,
+) -> float:
+    """Calculates rate of ATP production in a species.
+
+    Args:
+     R: Ribosome fraction
+     s: substrate concentration
+     w: waste product concentration
+     reaction_energy: the ATP generated for each species for each reaction alpha
+     v: the ith species' proportional expression level for reaction alpha
+     Gatp: ATP free energy
+    """
+    u = reaction_energy * q(R, s, w, reaction_energy, v)
+    return u
+
+
 def da(
-    ta: float, a: float, R: float, j: float = 2e5, chi: float = 29, m: float = 1e8
+    ta: float,
+    a: NDArray[np.float32],
+    R: NDArray[np.float32],
+    s: float,
+    w: float,
+    reaction_energy: NDArray[np.float32],
+    v: NDArray[np.float32],
+    chi: float = 29,
+    m: float = 1e8,
 ) -> float:
     """Rate of change of internal energy (ATP) production.
 
     Args:
     ta: time
     a: internal energy (ATP) concentration
     R: Ribosome fraction
-    j: ATP production rate
+    s: substrate concentration
+    w: waste product concentration
+    reaction_energy: the ATP generated for each species for each reaction alpha
+    v: the ith species' proportional expression level for reaction alpha
     chi: is ATP use per elongation step
     m: total mass of the cell (in units of amino acids)
     """
-    lam = calculate_lam(a, R)
-    uda = j - chi * m * lam - a * lam
-    return uda
+    a_changes = np.zeros(len(a))
+    for ind, _ in enumerate(a):
+        j = calculate_j(R[ind], s, w, v[ind], reaction_energy[ind])
+        lam = calculate_lam(a[ind], R[ind])
+        a_changes[ind] = j - chi * m * lam - a[ind] * lam
+    return a_changes
+
+
+def dc(
+    c: NDArray[np.float32],
+    reaction_energy: NDArray[np.float32],
+    N: NDArray[np.float32],
+    R: NDArray[np.float32],
+    v: NDArray[np.float32],
+    k: float = 3.3e-7,
+    p: float = 6e-5,
+) -> float:
+    """Calculates the change in metabolite concentration.
+
+    Args:
+     c: metabolite concentration
+     reaction_energy: the ATP generated for each species for each reaction alpha
+     N: current population
+     R: Ribosome fraction
+     v: the ith species' proportional expression level for reaction alpha
+     k: substrate supply rate
+     p: metabolite dilution rate
+    """
+    c_changes = np.zeros(len(c))
+    for ind, _ in enumerate(c):
+        if ind == 0:
+            c_changes[ind] = (
+                k
+                - p * c[ind]
+                - N[ind] * q(R[ind], c[0], c[1], reaction_energy[ind], v[ind])
+            )
+        else:
+            c_changes[ind] = -p * c[ind] + N[ind] * q(
+                R[ind], c[0], c[1], reaction_energy[ind], v[ind]
+            )
```

Comment:
Okay so the above should become
```python
for metabolite_num, _ in enumerate(c):
    for species_num, _ in enumerate(N):
        if metabolite_num == 0:
            c_changes[metabolite_num] += (
                k
                - p * c[metabolite_num]
                - N[species_num] * q(R[species_num], c[0], c[1], reaction_energy[species_num], v[species_num])
            )
        else:
            c_changes[metabolite_num] += -p * c[metabolite_num] + N[species_num] * q(
                R[species_num], c[0], c[1], reaction_energy[species_num], v[species_num]
            )
```

Reply:
Basically need to loop over all metabolites, but then for each metabolite calculate the contribution for each species (This is done by looping over each species)